### PR TITLE
[codex] add replay consistency harness

### PR DESCRIPTION
**Pull request by GodBlf**
## Summary

- Add `session/replaytest` replay consistency harness for Session / Memory / Summary / Track backends.
- Provide 10 public deterministic replay cases covering dialogue ordering, tools, state, memory, summaries, tracks, interleaving, and retry recovery.
- Add normalization, diff reporting, lightweight InMemory vs JSON-file persistent backend comparison, docs, and sample diff report.

Closes #2001

## Validation

- `go test -count=1 ./session/replaytest`
- `go test ./session/...`
- `gofmt -r 'interface{} -> any' -l session/replaytest`

### CODE REVIEW

File: session/replaytest/compare.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+			SessionID:    base.SessionID,
+			Locator:      locator,
+			FieldPath:    path,
+			BaseValue:    b,
+			CompareValue: c,
+			AllowedDiff:  false,
+			Explanation:  explanation,
+		})
+	}
+	add("$.session_id", "session", base.SessionID, compare.SessionID, "session ownership changed")
```
Comment (Flash-LHR) [inline session/replaytest/compare.go:84]: CompareSnapshots ignores `app_name` and `user_id`, so wrong session owners pass with no diff. Add both comparisons next to `session_id`.
<details><summary>中文</summary>
CompareSnapshots 忽略 `app_name` 和 `user_id`，错误的 session 归属会无 diff 通过。请在 `session_id` 旁边补充这两个比较。
</details>

File: session/replaytest/compare.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+func compareMaps[T any](path string, base, compare map[string]T, add func(string, string, any, any)) {
+	keys := map[string]struct{}{}
+	for k := range base {
+		keys[k] = struct{}{}
+	}
+	for k := range compare {
+		keys[k] = struct{}{}
+	}
+	for k := range keys {
```
Comment (Flash-LHR) [inline session/replaytest/compare.go]: compareMaps iterates map keys unsorted, so state diff order is nondeterministic. Sort keys before adding diffs.
<details><summary>中文</summary>
compareMaps 未排序遍历 map key，state diff 顺序不确定。请排序后再追加 diff。
</details>

**Comment by GodBlf** [comment]
I have read the CLA Document and I hereby sign the CLA

File: session/replaytest/backend.go [hunk 3]
```diff
@@ -0,0 +1,812 @@
...
+	}
+
+	if err := run.applyOperations(); err != nil {
+		return nil, err
+	}
+	got, err := sessions.GetSession(ctx, c.Key)
+	if err != nil {
+		return nil, err
+	}
+	run.stateOverlay.apply(got)
```
Comment (Rememorio) [inline session/replaytest/backend.go:95]: **P1: Do not patch state after reading the backend**

This overlay replaces the state returned by `GetSession` with the expected operation result. `CreateSession` returns a clone, and delete/clear only mutate that clone, so those operations are never persisted through the in-memory service; the overlay then hides that fact. It also masks a backend that returns success from state or event updates without actually storing the delta.

Normalize only the freshly read backend state. Delete and clear should be executed through a real backend operation, or explicitly reported as unsupported when the service contract cannot express them. A regression test should inspect a second raw `GetSession` after set/delete/clear rather than the patched snapshot.

<details><summary>中文</summary>

**不要在读取后修补后端 state**

这里会用操作的预期结果覆盖 `GetSession` 实际返回的 state。`CreateSession` 返回的是副本，而 delete/clear 目前只修改这个副本，并没有通过内存 session service 持久化；随后 overlay 又把这个问题隐藏了。若后端更新接口返回成功但没有真正保存 state delta，同样会被掩盖。

归一化时应只使用重新读取到的后端状态。delete/clear 需要通过真实的后端操作执行；如果公共 service 契约无法表达，则应明确标记为 unsupported。回归测试应在 set/delete/clear 后再次直接调用 `GetSession` 验证，而不是检查修补后的 snapshot。

</details>

File: session/replaytest/backend.go [hunk 4]
```diff
@@ -0,0 +1,812 @@
...
+		return nil
+	}
+	if _, ok := seen[spec.LogicalID]; ok {
+		return nil
+	}
+	evt, err := eventFromSpec(*spec)
+	if err != nil {
+		return err
+	}
+	seen[spec.LogicalID] = struct{}{}
```
Comment (Rememorio) [inline session/replaytest/backend.go:578]: **P1: The retry path never retries a failed append**

The event is marked as seen before `AppendEvent` completes. More importantly, `applyEventOperation` returns immediately on the first error, while a successful first append causes the second call to be skipped by this map. `OpRetryEvent` therefore only deduplicates two successful submissions; it never exercises the advertised failure-recovery behavior.

Add a deterministic fail-once append seam: the first call should fail without committing, the retry should succeed, and the final read should contain exactly one event and one applied state delta. Record deduplication state only according to the confirmed persistence outcome.

<details><summary>中文</summary>

**retry 路径不会重试失败的 append**

事件在 `AppendEvent` 完成前就被标记为 seen。更关键的是，第一次调用失败时 `applyEventOperation` 会立即返回；第一次成功时，第二次调用又会被 seen map 跳过。因此 `OpRetryEvent` 只是在去重两次成功提交，并没有覆盖所宣称的失败恢复语义。

建议增加一个确定性的 fail-once append 注入点：第一次调用失败且不落库，第二次重试成功，最终读取结果中只能有一个事件并且 state delta 只应用一次。去重状态也应根据已确认的持久化结果更新。

</details>

File: session/replaytest/compare.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+			)
+		}
+		report.Cases = append(report.Cases, caseReport)
+	}
+	return report, nil
+}
+
+// CompareSnapshots returns all normalized differences between base and compare.
+func CompareSnapshots(base, compare *Snapshot) []Difference {
+	if base == nil || compare == nil {
```
Comment (Rememorio) [inline session/replaytest/compare.go:64]: **P1: Reject nil snapshots instead of reporting no differences**

A backend that accidentally returns `(nil, nil)` is treated as having no differences. `Run` does not validate the returned pointer, so this produces a successful report even when either the reference backend or a compared backend supplied no result at all.

Have `Run` return an error naming the case and backend when `Apply` returns a nil snapshot, or emit an explicit blocking presence difference. Add a fake backend returning `(nil, nil)` to lock down the fail-closed behavior.

<details><summary>中文</summary>

**应拒绝 nil snapshot，而不是报告无差异**

如果后端意外返回 `(nil, nil)`，这里会把它当成“没有差异”。`Run` 也没有验证返回指针，因此无论基准后端还是被比较后端完全没有提供结果，都可能生成成功报告。

建议在 `Apply` 返回 nil snapshot 时，由 `Run` 返回包含 case 和 backend 名称的错误；或者生成明确的阻断性 presence diff。可增加一个返回 `(nil, nil)` 的 fake backend，确保框架采用失败关闭语义。

</details>

File: session/replaytest/normalize.go [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+			sort.Strings(participants)
+			metadata["participants"] = strings.Join(participants, ",")
+		}
+		if entry.Memory.Location != "" {
+			metadata["location"] = entry.Memory.Location
+		}
+		topics := append([]string{}, entry.Memory.Topics...)
+		sort.Strings(topics)
+		content := entry.Memory.Memory
+		stable := stableID(entry.AppName, entry.UserID, content, strings.Join(topics, ","), metadata["kind"])
```
Comment (Rememorio) [inline session/replaytest/normalize.go:134]: **P1: Align stable memory identity with episodic metadata semantics**

This stable ID includes topics but omits event time, participants, and location. The repository's `GenerateMemoryID` intentionally does the opposite for identity: topics may drift without changing identity, while episodic metadata distinguishes separate episodes with identical text.

As written, same-content episodes can receive identical stable IDs and then be ordered according to backend-specific read order, producing unstable or incorrectly paired diffs; a topic-only update is also presented as an identity change. Derive the normalized identity from the canonical memory identity fields and add cases for two same-content episodes with different metadata plus a topic-only update.

<details><summary>中文</summary>

**稳定 memory 身份应与 episodic metadata 语义一致**

这里的 stable ID 包含 topics，却遗漏了 event time、participants 和 location。仓库现有的 `GenerateMemoryID` 对身份语义有明确约定：topics 变化不应改变身份，而 episodic metadata 用于区分文本相同但实际不同的事件记忆。

当前实现会让同内容的不同 episode 得到相同 stable ID，后续排序依赖各后端的读取顺序，可能产生不稳定或错误配对的 diff；仅更新 topics 也会被表示为身份变化。建议按现有 canonical memory identity 字段生成归一化身份，并增加“同内容、不同 metadata 的两个 episode”以及“只更新 topics”的用例。

</details>

File: session/replaytest/backend.go [hunk 7]
```diff
@@ -0,0 +1,812 @@
...
+	}
+	return nil
+}
+
+func (r *inMemoryRun) applyOperation(op Operation) error {
+	switch op.Kind {
+	case OpAppendEvent, OpRetryEvent:
+		return r.applyEventOperation(op)
+	case OpSetState, OpDeleteState, OpClearState:
+		return r.applyStateOperation(op)
```
Comment (Rememorio) [inline session/replaytest/backend.go:132]: **P2: Fail closed on unknown replay operations**

The exported `ReplayCase` API accepts arbitrary `OperationKind` values, but this switch silently ignores anything it does not recognize. A typo, a serialized newer operation, or an adapter missing a newly added operation can therefore execute no work and still produce a clean consistency report.

Return an error from the default branch that includes the operation kind and case. Unsupported capabilities should continue through the explicit `OpUnsupportedProbe` path rather than being indistinguishable from an adapter omission; apply the same validation to the JSON-file adapter.

<details><summary>中文</summary>

**遇到未知回放操作时应失败关闭**

公开的 `ReplayCase` API 可以接收任意 `OperationKind`，但这个 switch 会静默忽略无法识别的值。这样一来，拼写错误、序列化的新版本操作，或适配器漏实现的新操作，都可能什么也不执行却仍然得到干净的一致性报告。

建议在 default 分支返回包含 operation kind 和 case 的错误。真正不支持的能力应继续通过明确的 `OpUnsupportedProbe` 路径报告，不应与适配器漏实现混为一谈；JSON-file 适配器也需要相同校验。

</details>

File: session/replaytest/backend.go [hunk 8]
```diff
@@ -0,0 +1,812 @@
...
+		return "jsonfile replay backend stores full event logs and does not implement strict page reads"
+	case CapabilityTTL:
+		return "jsonfile replay backend does not expire records during lightweight consistency tests"
+	default:
+		return "capability is not exposed by the jsonfile replay adapter"
+	}
+}
+
+func (b *jsonFileBackend) Apply(ctx context.Context, c ReplayCase) (*Snapshot, error) {
+	_ = ctx
```
Comment (Rememorio) [inline session/replaytest/backend.go:331]: **P2: Honor cancellation in the JSON-file backend**

This backend discards the context and then performs every replay operation and file rewrite regardless of cancellation. A canceled `Run` can therefore continue through a large custom case even though other backends use the same context for their storage calls.

Check `ctx.Err()` before initialization, before each operation, and before the final read, propagating the context error. A canceled-context test with multiple operations should assert that no later operations are applied.

<details><summary>中文</summary>

**JSON-file 后端应响应取消信号**

该后端完全忽略 context，取消之后仍会继续执行所有回放操作和文件重写。对于操作数量较多的自定义 case，即使同一次 `Run` 中其他后端已经响应取消，它仍可能继续工作。

建议在初始化前、每次操作前以及最终读取前检查 `ctx.Err()`，并原样返回 context 错误。可增加一个已取消 context 的多操作用例，断言后续操作不会再被应用。

</details>

File: session/replaytest/normalize.go [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+		sort.Strings(topics)
+		updated := "unset"
+		if !sum.UpdatedAt.IsZero() {
+			updated = "set"
+		}
+		out = append(out, NormalizedSummary{
+			FilterKey:      filterKey,
+			Text:           sum.Summary,
+			Version:        version,
+			SessionID:      sess.ID,
```
Comment (Rememorio) [inline session/replaytest/normalize.go:185]: **P2: Summary ownership is synthesized, not validated**

`SessionID` is copied from the containing session rather than from summary storage provenance. If a persistent backend mistakenly loads another session's summary row into this map, normalization still records the requested session ID, so the advertised ownership check cannot detect the regression. The current injected test mutates the snapshot after normalization and does not cover this path.

Capture the owner key while the adapter reads summary records and carry it into `NormalizedSummary`, or remove ownership from the claimed comparison when the backend cannot expose it. Add a fixture where the summary row belongs to a different session before normalization.

<details><summary>中文</summary>

**summary 归属是合成值，并未真正验证**

这里的 `SessionID` 直接取自外层 session，而不是 summary 在存储中的真实归属信息。如果持久化后端错误地把另一个 session 的 summary 行加载进当前 map，归一化结果仍会写成请求的 session ID，因此无法发现所宣称的归属错误。现有测试是在归一化后直接篡改 snapshot，没有覆盖真实读取路径。

建议由适配器在读取 summary 记录时保留 owner key，并写入 `NormalizedSummary`；如果后端无法暴露归属信息，则不应宣称能比较该字段。测试中应在归一化前构造一个归属于其他 session 的 summary 记录。

</details>

**Review by Rememorio — COMMENTED** [review]
I reviewed the changed lines and left 7 inline comments below. The comments focus on issues worth addressing before merge.

Reviewed the replay harness against existing session and memory contracts. The main risks are fail-open adapter behavior, post-read state patching that masks persistence regressions, retry coverage that never exercises failure recovery, and normalization that diverges from memory identity semantics. Add focused fake-backend tests for nil snapshots, unknown operations, failed-first appends, raw persisted state, and same-content episodic memories.

<details><summary>中文</summary>

我看过这次变更的相关代码，在下面留下 7 条行内评论。评论聚焦在合并前值得处理的问题。

已结合现有 session 和 memory 契约审查该回放测试框架。主要风险包括：适配器遇到异常输入时静默放行、读取后修补 state 从而掩盖持久化回归、retry 用例没有真正覆盖失败恢复，以及 memory 归一化身份规则与现有语义不一致。建议补充针对 nil snapshot、未知操作、首次 append 失败、原始持久化 state 和同内容 episodic memory 的定向测试。

</details>

**Comment by liuzengh** [comment]
感谢你对本 Issue 的贡献！

犀牛鸟活动的 Issue 提交阶段已经结束，本 PR 将关闭。再次感谢你的参与与付出！
